contrib/endpoints/src/api_manager/service_control/aggregated.cc

```diff
@@ -341,13 +439,79 @@ Status Aggregated::GetStatistics(Statistics* esp_stat) const {
   return Status::OK;
 }
 
+template <class RequestType>
+const std::string& Aggregated::GetApiReqeustUrl(const RequestType& request) {
+  if (typeid(request) == typeid(CheckRequest)) {
+    return url_.check_url();
+  } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+    return url_.quota_url();
+  } else {
+    return url_.report_url();
+  }
+}
+
+template <class RequestType>
+int Aggregated::GetHttpRequestTimeout(const RequestType& request) {
+  int timeout_ms = 0;
+
+  // Set timeout on the request if it was so configured.
+  if (typeid(request) == typeid(CheckRequest)) {
+    timeout_ms = kCheckDefaultTimeoutInMs;
+  } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+    timeout_ms = kAllocateQuotaDefaultTimeoutInMs;
+  } else {
+    timeout_ms = kReportDefaultTimeoutInMs;
+  }
+
+  if (server_config_ != nullptr &&
+      server_config_->has_service_control_config()) {
+    const auto& config = server_config_->service_control_config();
+    if (typeid(request) == typeid(CheckRequest)) {
+      if (config.check_timeout_ms() > 0) {
+        timeout_ms = config.check_timeout_ms();
+      }
+    } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+      if (config.quota_timeout_ms() > 0) {
+        timeout_ms = config.quota_timeout_ms();
+      }
+    } else {
+      if (config.report_timeout_ms() > 0) {
+        timeout_ms = config.report_timeout_ms();
+      }
+    }
+  }
+
+  return timeout_ms;
+}
+
+template <class RequestType>
+const std::string& Aggregated::GetAuthToken(const RequestType& request) {
+  if (sa_token_) {
+    if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+      return sa_token_->GetAuthToken(
+          auth::ServiceAccountToken::JWT_TOKEN_FOR_QUOTA_CONTROL);
+    } else {
+      return sa_token_->GetAuthToken(
+          auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICE_CONTROL);
+    }
+  } else {
+    static std::string empty;
+    return empty;
+  }
+}
+
 template <class RequestType, class ResponseType>
 void Aggregated::Call(const RequestType& request, ResponseType* response,
                       TransportDoneFunc on_done,
                       cloud_trace::CloudTraceSpan* parent_span) {
   std::shared_ptr<cloud_trace::CloudTraceSpan> trace_span(
       CreateChildSpan(parent_span, "Call ServiceControl server"));
-  std::unique_ptr<HTTPRequest> http_request(new HTTPRequest([response, on_done,
+
+  const std::string& url = GetApiReqeustUrl(request);
+  TRACE(trace_span) << "Http request URL: " << url;
```

Comment:
I have tried. I might be tested wrong way. Please correct me.

in aggregated.h

  template <class RequestType>
  const std::string& GetApiReqeustUrl();

in aggregated.cc

template <class RequestType>
const std::string& Aggregated::GetApiReqeustUrl() {
  if (typeid(RequestType) == typeid(CheckRequest)) {
    return url_.check_url();
  } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
    return url_.quota_url();
  } else {
    return url_.report_url();
  }
}

in Aggregated::Call()

  const std::string& url = GetApiReqeustUrl();


$ bazel build //contrib/endpoints/src/api_manager:api_manager
INFO: Found 1 target...
ERROR: /usr/local/google/home/jaebong/workspace/proxy/contrib/endpoints/src/api_manager/service_control/BUILD:19:1: C++ compilation of rule '//contrib/endpoints/src/api_manager/service_control:service_control' failed: linux-sandbox failed: error executing command /usr/local/google/home/jaebong/.cache/bazel/_bazel_jaebong/944d8bc865545967743040c9d7fb6855/execroot/proxy/_bin/linux-sandbox ... (remaining 91 argument(s) skipped).
contrib/endpoints/src/api_manager/service_control/aggregated.cc: In member function 'void google::api_manager::service_control::Aggregated::Call(const RequestType&, ResponseType*, google::service_control_client::TransportDoneFunc, google::api_manager::cloud_trace::CloudTraceSpan*)':
contrib/endpoints/src/api_manager/service_control/aggregated.cc:529:45: error: no matching function for call to 'google::api_manager::service_control::Aggregated::GetApiReqeustUrl()'
   const std::string& url = GetApiReqeustUrl();
                                             ^
contrib/endpoints/src/api_manager/service_control/aggregated.cc:529:45: note: candidate is:
In file included from contrib/endpoints/src/api_manager/service_control/aggregated.cc:17:0:
./contrib/endpoints/src/api_manager/service_control/aggregated.h:120:22: note: template<class RequestType> const string& google::api_manager::service_control::Aggregated::GetApiReqeustUrl()
   const std::string& GetApiReqeustUrl();
                      ^
./contrib/endpoints/src/api_manager/service_control/aggregated.h:120:22: note:   template argument deduction/substitution failed:
contrib/endpoints/src/api_manager/service_control/aggregated.cc:529:45: note:   couldn't deduce template parameter 'RequestType'
   const std::string& url = GetApiReqeustUrl();
                                             ^
Target //contrib/endpoints/src/api_manager:api_manager failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 3.033s, Critical Path: 2.83s